### DISCUSSION TITLE
Filter / Lua API: access Quarto's embedded LaTeX distribution?

### DISCUSSION THREAD

**Discussion by jdutant** [quarto-dev/quarto-cli#11543] [Q&A] [answered]
### Description

Is there a way for a Lua filter to execute commands from Quarto's own tinytex distribution? 

The use case I have in mind is this: the [Imagify filter](https://github.com/dialoa/imagify) converts LaTeX/TikZ elements into images when targeting non-PDF formats using `latexmk` and `dvisvgm` (part of TeXLive). If the filter could make use of Quarto's tinytex distribution and, if needed, install dvisvgm on the fly, it would work out of the box. As it stands, people need to have a separate LaTeX installation.

**Comment by mcanouil** [comment]
I think there is a slight misunderstanding here.
Quarto does not embed a LaTeX distribution.
Quarto allows you to install TinyTex (it provides an helper command for that) and use that installation (if performed by the user).
If the distribution is added to the PATH, then you have access to the tools from everywhere, but the install and the fact that it is added or not to the PATH is up to the user.

Additionally, all the logic to find available LaTeX distribution is done via TypeScript so you won't have access to that install LUA.

**Reply by cderv** [reply]
When a user does `quarto install tinytex` it won't be added to PATH by default 🤔 So we could imaging pass the information about the path to this tinytex install to Lua, or even the engine path that will be used by Quarto - I think we do it for `Rscript` path used by Quarto. 

This would ensure using the same version. But for LaTeX one usually has one distribution, so for the filter requirement the latex distribution to be in PATH seems a right requirement. 

@jdutant you would also have your filter look in PATH, or read an environment variable or a config that a user could set if one wants to specify a path. 
I don't think quarto is currently limiting your idea, unless I missing something.

**Reply by jdutant** [reply]
Thanks both, that's helpful. I thought Quarto kept TinyTeX in its folders because the docs call it Quarto's "internal version" of LaTeX ([here](https://quarto.org/docs/output-formats/pdf-engine.html#installing-tex)) and call latexmk "Quarto's built-in compilation engine" ([here](https://quarto.org/docs/output-formats/pdf-engine.html#quarto-pdf-engine)). I just tried `quarto install tinytex`, and it looks like it's either running [TinyTeX's install script](https://yihui.org/tinytex/) or at least placing TinyTeX in the same default folders. I guess the filter could look them up if there's no latexmk in PATH. 

> But for LaTeX, one usually has one distribution, so for the filter requirement, the latex distribution to be in PATH seems a right requirement.

I agree, except for the case in which a user installs tinytex without `--update-path` and doesn't have a LaTeX installation otherwise. Then Quarto is able to produce PDFs but the filter couldn't yet produce the corresponding HTML. I think it'd be a nice default for the filter to use whatever Quarto uses unless otherwise specified. 

Thanks!

**Reply by cderv** [reply]
> I just tried quarto install tinytex, and it looks like it's either running [TinyTeX's install script](https://yihui.org/tinytex/) or at least placing TinyTeX in the same default folders. I guess the filter could look them up if there's no latexmk in PATH.

Yes it will install at usual place at user level that other way to install TinyTeX. 

> call latexmk "Quarto's built-in compilation engine" ([here](https://quarto.org/docs/output-formats/pdf-engine.html#quarto-pdf-engine))

Quarto default engine is to use its own latexmk emulation. It won't call directly `latexmk` unless you opt out quarto's engine and set it as a pdf engine as explain in the doc. The quarto compilation mention is there to automatically detect errors and try install missing package, and also re-run as necessary to solve index and bibliography entry. 

> I agree, except for the case in which a user installs tinytex without --update-path and doesn't have a LaTeX installation otherwise.

Yes exactly. You would need to make it a requirement for your filter to be in PATH, or you could look at same place as quarto. I'll see how / if I can expose the path found by quarto to filters like we do with R script. 🤔

**Reply by cderv** [reply]
>  I'll see how / if I can expose the path found by quarto to filters like we do with R script. 🤔

I opened 
- https://github.com/quarto-dev/quarto-cli/pull/11670

**Reply by cderv** [reply, marked as answer]
`quarto.paths.tinytex_bin_dir()` is now exposed in Quarto Lua API to know where Quarto found TinyTeX if installed. Otherwise, it means that Quarto has found latex tools in PATH. 
See https://prerelease.quarto.org/docs/extensions/lua-api.html#paths

**Reply by jdutant** [reply]
Sweet!
